File: 32. Longest Valid Parentheses/optimised.ts
function longestValidParentheses(s: string): number {
	const n = s.length;

	const slack: number[] = [-1];
	let max_len = 0;
	for (let i = 0; i < n; i++) {
		if (s[i] === '(') {
			slack.push(i);
		} else {
			slack.pop();

			if (slack.length === 0) {
				slack.push(i);
			} else {
				max_len = Math.max(max_len, i - slack[slack.length - 1]);
			}
		}
	}

	return max_len;
};

console.log(longestValidParentheses("(()"));
console.log(longestValidParentheses(")()())"));
console.log(longestValidParentheses(""));

File: 32. Longest Valid Parentheses/optimised.py
#!/usr/bin/env python

def longestValidParentheses(s: str) -> int:
    n = len(s)
    stack = [-1]
    max_len = 0
    for i in range(n):
        if s[i]  == '(':
            stack.append(i)
        else:
            stack.pop()

            if not stack:
                stack.append(i)
            else:
                l = i - stack[-1]
                max_len = max(max_len, l)

    return max_len

print(longestValidParentheses("(()"))
print(longestValidParentheses(")()())"))
print(longestValidParentheses(""))
